packages/@ember/-internals/runtime/lib/mixins/array.d.ts

```diff
@@ -0,0 +1,103 @@
+    callback: (this: Target, item: T, index: number, arr: this) => void,
+    target?: Target
+  ): this;
+  getEach<K extends string>(key: K): NatifiedArray<Value<T, K>>;
+  setEach<K extends string>(key: K, value: Value<T, K>): this;
+  map<R, Target>(
+    callback: (this: Target, item: T, index: number, arr: this) => R,
+    target?: Target
+  ): NatifiedArray<R>;
+  mapBy<K extends string>(key: K): NatifiedArray<Value<T, K>>;
+  filter<Target>(
+    callback: (this: Target, item: T, index: number, arr: this) => boolean,
+    target?: Target
+  ): NatifiedArray<T>;
+  reject<Target>(
+    callback: (this: Target, item: T, index: number, arr: this) => boolean,
+    target?: Target
+  ): NatifiedArray<T>;
+  filterBy(key: string): NatifiedArray<T>;
+  rejectBy(key: string): NatifiedArray<T>;
+  find<Target = null>(
+    callback: (this: Target, item: T, index: number, arr: this) => boolean,
+    target?: Target
+  ): T | undefined;
+  findBy<K extends string>(key: K, value?: Value<T, K>): T | undefined;
+  every<Target = null>(
+    callback: (this: Target, item: T, index: number, arr: this) => boolean,
+    target?: Target
+  ): boolean;
+  isEvery<K extends string>(key: K, value?: Value<T, K>): boolean;
+  any<Target = null>(
+    callback: (this: Target, item: T, index: number, arr: this) => boolean,
+    target?: Target
+  ): boolean;
+  isAny<K extends string>(key: K, value?: Value<T, K>): boolean;
+  reduce<V>(
+    callback: (summation: V, current: T, index: number, arr: this) => V,
+    initialValue?: V
+  ): V;
+  invoke<K extends string>(
+    methodName: K,
+    ...args: Value<T, K> extends (...args: any[]) => any ? Parameters<Value<T, K>> : unknown[]
+  ): NatifiedArray<Value<T, K> extends (...args: any[]) => any ? ReturnType<Value<T, K>> : unknown>;
+  toArray(): T[];
+  compact(): NatifiedArray<Exclude<T, null>>;
+  includes(object: T, startAt?: number): boolean;
+  sortBy(key: string): T[];
+  uniq(): NatifiedArray<T>;
+  uniqBy(key: string): NatifiedArray<T>;
+  without(value: T): NatifiedArray<T>;
+}
+
+export default EmberArray;
+
+interface MutableArray<T> extends EmberArray<T> {
+  replace(idx: number, amt: number, objects: T[]): void;
+}
+
+declare class MutableArray<T> extends EmberArray<T> {
+  clear(): this;
+  insertAt(idx: number, object: T): this;
+  removeAt(start: number, len: number): this;
+  pushObject(obj: T): this;
+  pushObjects(objects: T[]): this;
+  popObject(): T | undefined;
+  shiftObject(): T | null | undefined;
+  unshiftObject(object: T): this;
+  unshiftObjects(objects: T[]): this;
+  reverseObjects(): this;
+  setObjects(object: T[]): this;
+  removeObject(object: T): this;
+  removeObjects(objects: T[]): this;
+  addObject(obj: T): this;
+  addObjects(objects: T[]): this;
+}
+
+export { MutableArray };
+
+export type NativeArray<T> = Array<T> & MutableArray<T>;
+
+export function A<T>(
+  arr?: T
+): T extends Array<infer TT> ? TT[] & NativeArray<TT> : unknown[] & NativeArray<unknown>;
```

Comment:
Two notes:

- the `TT` is a bit unusual in my experience; maybe `A` instead?
- Given how `NativeArray` is defined, I think the intersections are redundant

```suggestion
): T extends Array<infer A> ? NativeArray<A> : NativeArray<unknown>;
```

Reply:
Yes, this may not be necessary if we get NativeArray defined correctly.